app/gui2/src/providers/widgetRegistry.ts

```diff
@@ -115,16 +206,45 @@ function isWidgetModule(module: unknown): module is WidgetModule {
   )
 }
 
-function isWidgetComponent(component: unknown): component is WidgetComponent {
+function isWidgetComponent(component: unknown): component is WidgetComponent<any> {
   return typeof component === 'object' && component !== null && 'render' in component
 }
 
-function isWidgetDefinition(config: unknown): config is WidgetDefinition {
+function isWidgetDefinition(config: unknown): config is WidgetDefinition<any> {
   return typeof config === 'object' && config !== null && 'priority' in config && 'match' in config
 }
 
-export function defineWidget(definition: WidgetDefinition): WidgetDefinition {
-  return definition
+/**
+ *
+ * @param match Filter the widget input type to only accept specific types of input. The declared
+ * widget props will depend on the type of input accepted by this filter. Only widgets that pass
+ * this filter will be scored and potentially used.
+ */
+export function defineWidget<M extends InputMatchFn<any> | InputMatchFn<any>[]>(
```

Comment:
as mentioned in other places - is `<T>` + `match: InputMatchFn<T> | InputMatchFn<T>[]` not preferable here?

Reply:
Unfortunately not. This messes up the inference direction, and things gets too easily coerced into whole `WidgetInput` type. Alternatively, it is not possible to specify an array of predicates that match different types (which is the whole point).